pyccel/ast/builtins.py

```diff
@@ -774,6 +775,40 @@ def is_homogeneous(self):
         """
         return True
 
+
+class PythonSetFunction(PyccelFunction):
+    """
+    Class representing a call to the `set` function.
+
+    Class representing a call to the `set` function. This is
+    different to the `{,}` syntax as it only takes one argument
+    and unpacks any variables.
+
+    Parameters
+    ----------
+    arg : TypedAstNode
+        The argument passed to the function call.
+    """
+
+    __slots__ = ('_shape', '_class_type')
+    name = 'set'
+    def __new__(cls, arg):
+        if isinstance(arg.class_type, HomogeneousSetType):
+            return arg
+        elif isinstance(arg, PythonList):
+            return PythonSet(*arg)
+        elif isinstance(arg.shape[0], LiteralInteger):
+            return PythonSet(*[arg[i] for i in range(arg.shape[0])])
+        elif isinstance(arg.shape[0], PyccelArrayShapeElement):
+            return super().__new__(cls)
+        else:
+            raise TypeError(f"Can't unpack {arg} into a Set")
```

Comment:
Why do you require the shape to be a `PyccelArrayShapeElement`?
I am concerned that this may exclude the following code (or may if we modify slightly how shapes are changed):
```python
def f():
    n : 'const int' = 5
    a : 'const float[:]' = np.ones(n+1)
```
And I don't see a justification for this

Reply:
I used `PyccelArrayShapeElement` to handle `an argument (unknown length)`.

Reply:
What do you mean by "unknown length"? In the example above the length of `a` is `n+1`. Is this a known length?

Reply:
I was talking about the reason for using `PyccelArrayShapeElement`, Like this example : 
```python
def f(a : list[int]):
    d = set(a)
```
Can you please explain to me the logic behind the set and the example that you provided.

Reply:
The more complete example is:
```python
def f(n : 'const int'):
    a : 'const float[:]' = np.ones(n+1)
    b = set(a)
```
historically in this case the shape of `a` would have been `(PyccelAdd(Variable(n), LiteralInteger(1)),)`. I suspect that currently the shape would simply be `(PyccelArrayShapeElement('a',0),)` but at some point I would like to go back to the more explicit shape.

You said that "I used PyccelArrayShapeElement to handle an argument (unknown length).". This implies that a `PyccelArrayShapeElement` is the only way to describe an "unknown length". Hence my question
> What do you mean by "unknown length"? In the example above the length of a is n+1. Is this a known length?

Reply:
Like I mentioned before, the reason I used `PyccelArrayShapeElement` is for the example that I provided, which explains the `unknown length`. If you have a better solution, I would appreciate it.

Reply:
Fixed : 28479a7